libs/gl-plugin/src/tramp.rs

```diff
@@ -488,15 +454,99 @@ async fn reestablished_channels(
             Ok(_amount) => Some(channel_data),
             _ => None,
         })
-        .collect::<Vec<ChannelData>>())
+        .collect::<Vec<Channel>>())
 }
 
-struct ChannelData {
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct Channel {
     short_channel_id: cln_rpc::primitives::ShortChannelId,
     spendable_msat: u64,
     min_htlc_out_msat: u64,
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct ChannelContribution<'a> {
+    channel: &'a Channel,
+    contrib_msat: u64,
+}
+
+/// Finds an allocation that covers the target amount using as few channels as
+/// possible.
+fn find_minimal_allocation<'a>(
+    channels: &'a [Channel],
+    target_msat: u64,
+) -> Option<Vec<ChannelContribution<'a>>> {
+    // We can not allocate channels for a zero amount.
+    if target_msat == 0 {
+        return None;
+    }
+
+    // We'll use recursion with backtracking.
+    // - `idx` is the current channel index.
+    // - `target_msat` is the remaining amount.
+    // - `current` collects the allocations so far.
+    // - `best` stores the best solution found.
+    fn rec<'a>(
+        channels: &'a [Channel],
+        idx: usize,
+        target_msat: u64,
+        current: &mut Vec<ChannelContribution<'a>>,
+        best: &mut Option<Vec<ChannelContribution<'a>>>,
+    ) {
+        // Base case: If we've exactly allocated the target, record the solution.
+        if target_msat == 0 {
+            if best.is_none() || current.len() < best.as_ref().unwrap().len() {
+                *best = Some(current.clone())
+            }
+            return;
+        }
+
+        // Base case: We have processed all channels and could not find a
+        // solution.
+        if idx >= channels.len() {
+            return;
+        }
+
+        // Option 1: Skip the current channel.
+        rec(channels, idx + 1, target_msat, current, best);
+
+        // Option 2: Use the current channel.
+        let ch = &channels[idx];
+
+        // Each channel has an upper and a lower bound defined by the minimum
+        // htlc amount and the spendable amount.
+        let lower = ch.min_htlc_out_msat;
+        let upper = ch.spendable_msat.min(target_msat);
+
+        // If the remaining target amount is below the channel's lower bound we
+        // can not use it.
+        if target_msat < lower {
+            return;
+        }
+
+        // Try allocations from the upper bound down to the lower bound to
+        // maximize channel usage.
+        for alloc in (lower..=upper).rev() {
```

Comment:
This will iterate a lot? min_msat being 1 and amount being 1000 sat (1M msat) will iterate 1M times per channel? So 1M * 1M for 2 channels?

What if (thinking out loud here) you would start with the maximum per channel. Then if that doesn't work, try to leave the minimum minimum (so the lowest min_htlc_out_msat of the remaining channels), if that doesn't work leave the second lowest minimum, etc. So the allocation of the current channel would become min(spendable_msat, target_msat - min_htlc_out_msat of the next channel).

Reply:
Yes that will iterate a lot and I like your improvement! What if we would, instead of choosing the lowest `min_htlc_out_msat`, choose the `min_htlc_out_msat` of the second smallest channel to try to drain as many channels as possible to  make them available for re-allocation?

Reply:
That sounds ok. Do add a fallback to the third smallest channel, then the fourth smallest, etc, if it doesn't work.

Reply:
Sure, we want to recursively try bigger channels until we run out of options.